### PR TITLE
Use postMessage instead of directly making API calls in the overlay iframe.

### LINKED ISSUE #13406: Error: You can't access this resource as it requires 'view' access for the website id = 60.
I'm using Page Overlay for one of our sites and have super user permissions for my user. Therefore I don't understand, why I can't see these bubbles on the page overlay. Everthing else (like Open Row Evolution, Open Transitions, Open segmented Visitor Log) works fine.

Both sites (Matomo + Website) use TLS. But I only get the page, this error and after clicking ok in the right corner: "Loading following pages ..."

I enabled debug mode and got this output:
<pre>
DEBUG Piwik\Plugin\Manager[2018-09-10 15:33:36 UTC] [43c7b] Loaded plugins: CorePluginsAdmin, CoreAdminHome, CoreHome, WebsiteMeasurable, Diagnostics, CoreVisualizations, Proxy, API, Widgetize, Transitions, LanguagesManager, Actions, Dashboard, MultiSites, Referrers, UserLa
nguage, DevicesDetection, Goals, Ecommerce, SEO, Events, UserCountry, GeoIp2, VisitsSummary, VisitFrequency, VisitTime, VisitorInterest, RssWidget, Monolog, Login, UsersManager, SitesManager, Installation, CoreUpdater, CoreConsole, ScheduledReports, UserCountryMap, Live, C
ustomVariables, PrivacyManager, ImageGraph, Overlay, SegmentEditor, Insights, Morpheus, Contents, Resolution, DevicePlugins, Heartbeat, Intl, Marketplace, ProfessionalServices, UserId, CustomPiwikJs, DBStats, Provider, Bandwidth, CustomOptOut, LogViewer, SecurityInfo, Site
Migration, TasksTimetable                                                                                                                                                                                                                                                        
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] /var/www/matomo/core/Access.php(505): You can't access this resource as it requires 'view' access for the website id = 60.                                                                                                            
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #0 /var/www/matomo/core/Piwik.php(511): Piwik\Access->checkUserHasViewAccess('60')                                                                                                                                                    
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #1 /var/www/matomo/plugins/SitesManager/API.php(223): Piwik\Piwik::checkUserHasViewAccess('60')                                                                                                                                       
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #2 /var/www/matomo/plugins/Overlay/API.php(47): Piwik\Plugins\SitesManager\API->getSiteFromId('60')                                                                                                                                   
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #3 [internal function]: Piwik\Plugins\Overlay\API->getExcludedQueryParameters('60')                                                                                                                                                   
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #4 /var/www/matomo/core/API/Proxy.php(236): call_user_func_array(Array, Array)                                                                                                                                                        
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #5 /var/www/matomo/core/API/Request.php(262): Piwik\API\Proxy->call('\\Piwik\\Plugins\\...', 'getExcludedQuer...', Array)                                                                                                             
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #6 /var/www/matomo/plugins/API/Controller.php(41): Piwik\API\Request->process()                                                                                                                                                       
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #7 [internal function]: Piwik\Plugins\API\Controller->index()                                                                                                                                                                         
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #8 /var/www/matomo/core/FrontController.php(560): call_user_func_array(Array, Array)                                                                                                                                                  
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #9 /var/www/matomo/core/FrontController.php(146): Piwik\FrontController->doDispatch(NULL, NULL, NULL)                                                                                                                                 
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #10 /var/www/matomo/core/dispatch.php(34): Piwik\FrontController->dispatch()                                                                                                                                                          
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #11 /var/www/matomo/index.php(27):require_once('/var/www/matomo...')                                                                                                                                                                  
DEBUG API[2018-09-10 15:33:36 UTC] [43c7b] #12 {main}  
</pre>

I'm quite lost right now. Can somebody help me? :)

### PR DESCRIPTION
**Pull request by diosmosis**
This fix replaces https://github.com/matomo-org/matomo/pull/13420 as we noticed some other issues in Overlay when reviewing it.

Fixes #13406

### CODE REVIEW

File: plugins/Overlay/javascripts/Piwik_Overlay.js [hunk 1]
```diff
@@ -139,6 +144,7 @@ var Piwik_Overlay = (function () {
     function hashChangeCallback(urlHash) {
         var location = getOverlayLocationFromHash(urlHash);
         location = Overlay_Helper.decodeFrameUrl(location);
+        iframeOrigin = location.match(DOMAIN_PARSE_REGEX)[0];
```
Comment (tsteur) [inline plugins/Overlay/javascripts/Piwik_Overlay.js]: can we be sure there will be an entry with index `[0]`?

File: plugins/Overlay/javascripts/Piwik_Overlay.js [hunk 2]
```diff
@@ -158,6 +164,41 @@ var Piwik_Overlay = (function () {
...
+            var strData = event.data.split(':', 3);
+            if (strData[0] !== 'overlay.call') {
+                return;
+            }
+
+            var requestId = strData[1];
+            var url = decodeURIComponent(strData[2]);
+
+            var params = broadcast.getValuesFromUrl(url);
+
```
Comment (tsteur) [inline plugins/Overlay/javascripts/Piwik_Overlay.js]: Can we here only allow  specific API methods needed for overlay to increase the security?

File: plugins/Overlay/javascripts/Piwik_Overlay.js [hunk 3]
```diff
@@ -26,7 +29,9 @@ var Piwik_Overlay = (function () {
         $location.html('&nbsp;').unbind('mouseenter').unbind('mouseleave');
 
         iframeCurrentPage = currentUrl;
-        iframeDomain = currentUrl.match(/http(s)?:\/\/(www\.)?([^\/]*)/i)[3];
+        var m = currentUrl.match(DOMAIN_PARSE_REGEX);
+        iframeDomain = m[3];
+        iframeOrigin = m[0];
```
Comment (tsteur) [inline plugins/Overlay/javascripts/Piwik_Overlay.js]: I presume an attacker cannot set a different iframeOrigin to her or his liking? eg by crafting a url that matches the regex etc... Also `m[0]` and `m[3]` might not be defined for whatever reason... might need to add some check there to make sure they are set...

I'm thinking would it be possible for additional security to only accept origins of configured siteURLs? I think this check is done so far maybe in the twig but not sure if in JS

**Comment by diosmosis** [comment]
@tsteur Updated.

**Pull request merged**